Refactoring authentication with security

Nitoriubar opened the pull request:
## 📝 세부사항
OAuth 2.0 Social Login 기능 고도화

- `TokenProvider`
    - 헷갈리는 부분인 Token 생성 메서드 이름 변경(`createAuthToken` -> `create Access/Refresh Token`) 
    - 생성자의 매개변수(`key`, `refreshKey`)를 비밀키로 즉시 변경하여 Token을 생성해주는 방식으로 변경함
    - `create Access/Refresh Token`메서드의 토큰 유효기간을 파라미터로 넘겨 받지 않고 Token생성시 파라미터로 넣어주는 방식으로 변경함
- `AuthToken`
    - 생성자의 매개변수(`expiry`)의 타입 변경함(Date -> Long)
    - `createAuthToken`메서드에 토큰 유효기간이 설정해놓은대로 동작하는 지 확인하고자 Token생성 날짜 추가함
- `AuthService`
    - `AuthController`에서 처리하던 Token관련 로직들을 `AuthService`를 추가하여 변경함
- `RefreshTokenService`
    - `RefreshToken`이 존재하는 경우와 존재하지 않는 경우를 나누고 도메인 구조에 맞도록 `RefreshTokenService`를 생성함
- `OAuth2AuthenticationSuccessHandler`
    -  `onAuthenticationSuccess`를 하는 경우 어떻게 동작하는 지 한눈에 알기 쉽도록 `getTargetUrl`메서드와 `createTokenAndCookie`메서드로 나눠 기능 분리함
    - `determineTargetUrl`이 메서드의 이름에 맞는 것만 할 수 있도록 변경함

## 📗 참고사항
[OAuth2.0 & JWT 로그인](https://deeplify.dev/back-end/spring/oauth2-social-login#userprincipal)
[JWT Token 확인](https://jwt.io/)
[OIDC](https://betterprogramming.pub/the-complete-guide-to-oauth-2-0-and-openid-connect-protocols-35ebc1cbc11a)
[CORS](https://inpa.tistory.com/entry/WEB-%F0%9F%93%9A-CORS-%F0%9F%92%AF-%EC%A0%95%EB%A6%AC-%ED%95%B4%EA%B2%B0-%EB%B0%A9%EB%B2%95-%F0%9F%91%8F)
[CSRF Token](https://codevang.tistory.com/282)
[Consider defining a bean of type 'Service' in your configuration. 에러](https://mimah.tistory.com/entry/Spring-Consider-defining-a-bean-of-type-Service-in-your-configuration)
[JWT 설치 에러](https://kyung-a.tistory.com/32)

## 📎 관련 이슈
#13

Nitoriubar committed:
refactor : 로그인 기능

Nitoriubar committed:
refactor : 로그인 기능 고도화

- 전반적인 로그인 기능을 고도화 하였음.
- Refresh Token 추가 예정
- Refresh Token 로직 구상

Nitoriubar committed:
refactor : OAuth2.0 Social Login Test 완료

- 생성한 Access Token으로 내 프로필 정보 가져오기 진행함
- @PreAuthorize를 사용하여 User의 Role에 따라 접근 권한을 설정함
- 접근 권한이 없는 경우에 관한 Error Handling이 필요함

Nitoriubar committed:
refactor : OAuth2.0 Social Login 기능 TEST

- 생성한 Access Token을 Postman으로 내 프로필 정보 가져오기 진행힘
- @PreAuthorize를 사용하여 User의 Role에 따라 접근 권한을 적용함
- 접근 권한이 없는 경우 Error Handling이 필요함
- refactoring전의 로그인 기능 구현한 코드 정리함

Nitoriubar committed:
refactor : JWT Token을 발급방식 변경

- TokenSecret값을 저장하는 곳 변경
- TokenProvider 별도 생성

Nitoriubar committed:
feat : Refresh Token 생성과 재발급 구현

- Refresh Token과 Access Token을 발급하는 Key를 다르게 설정해줌
- Refresh Token의 유효기간이 3일 미만으로 남아있는 경우, 새로 생성해줌
- Refresh Token의 유효기간이 남아있고 Access Token은 만료된 경우 새로 Access Token발금해줌

Nitoriubar committed:
refactor : AuthController와 AuthService로 분리함

Nitoriubar committed:
refact : Token이 생성되는 부분

- TokenProvider에서 헷갈리는 메서드 이름 & 파라미터 변경함
- AuthToken에서 Token의 유효기간 설정을 바로 해주는 방식으로 변경 & 생성된 시간 추가
- OAuth2AuthenticationSuccessHandler에서 onAuthenticationSuccess를 하면 어떤 로직이 일어나는 지 한번에 보기 위한 로직으로 변경 & determineTargetUrl이 하는 기능만 남기고 나머지 부분은 메서드로 분리함
-OAuth2AuthenticationSuccessHandler에서 순환참조가 발생하여 RefreshTokenService를 생성하여 해결함

Nitoriubar committed:
style : 도메인 구조에 맞게 디렉터리 구조 변경

sangrimlee commented on src/main/java/com/triprint/backend/domain/user/entity/User.java:
```diff
@@ -88,12 +94,13 @@ public class User {
-                String kakaoEmail, UserRole userRole) {
+                String kakaoEmail, UserRole userRole, AuthProvider provider) {
 
         this.kakaoId = kakaoId;
```
이제 `User` Entity에서 `kakaoId`가 아니라 `providerId`라는 이름으로 관리하는 게 더 좋을 것 같아요!

Nitoriubar replied:
넵!! 수정했습니다.

sangrimlee commented on src/main/resources/application.yml:
```diff
@@ -61,8 +61,22 @@ cloud:
-token:
-  secret-key: ${TOKEN_SECRET}
-  expiration-time: ${EXPIRATION_TIME}
+jwt:
+  secret: ${JWT_SECRET}
```
이제는 사용하지 않는 걸로 알고 있는데 이 부분 삭제해주시면 될 것 같아요 :)

Nitoriubar replied:
네, 알겠습니다!

sangrimlee reviewed (approved):
고생하셨습니다👍

블로그글 보고 차용하신 `ResourceNotFoundException`과 같은 Exception, ApiResponse들을 이번에 수정하시면 될 것 같아요.

Nitoriubar committed:
feat : 전반적인 Error Handling 기능 구현

- 전반적인 'Error Handling'을 하기위한 틀로 'ServiceException'을 만들고, 세부적으로 'Handling'이 필요한 'Exception'들은 상속하여 구현함
    - 'ForbiddenException'
    - 'ResourceNotFoundException'
    - 'BadRequestException'
- 'GlobalExceptionHandler'에서 'Custom'한 Exception은 'ServiceException'클래스로 설정하고, 그 외 나머지 'Exception'은 'Exception'클래스로 설정하여 예외처리함

Nitoriubar committed:
refactor : Exception 마무리 & Double checking 완료
